### PR TITLE
replace golang.org/x/tools with golang.org/x/tools/go/vcs@v0.1.0-deprecated

### LINKED ISSUE #1579: golang.org/x/tools removed vcs package
### What version of gazelle are you using?

latest main

### What version of rules_go are you using?

latest (v0.41.0)

### What version of Bazel are you using?

6.2.1

### Does this issue reproduce with the latest releases of all the above?

yes

### What operating system and processor architecture are you using?

Ubuntu 22.04 / x86_64

### What did you do?

I tried to upgrade `golang.org/x/tools` in my repo. My repo also uses bazel-gazelle.

### What did you expect to see?

The upgrade should suceed.

### What did you see instead?

Gazelle does not compile anymore when I use it in my repo. Since it depends on `golang.org/x/tools/go/vcs`, which [has been removed in version `v0.11.0`](https://cs.opensource.google/go/x/tools/+/5a89a3bf267ef12790327b8692c88654845bc78d), minimum version selection will lead to a missing package:

```
INFO: Invocation ID: 70b7838e-b211-4814-8453-c488d6156b94
ERROR: /home/malte/.cache/bazel/_bazel_malte/bb1175a64d2f0f71785aea697a0484ea/external/bazel_gazelle/repo/BUILD.bazel:3:11: no such package '@org_golang_x_tools//go/vcs': BUILD file not found in directory 'go/vcs' of external repository @org_golang_x_tools. Add a BUILD file to a directory to mark it as a package. and referenced by '@bazel_gazelle//repo:repo'
ERROR: Analysis of target '//:tidy' failed; build aborted: 
INFO: Elapsed time: 0.236s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded, 0 targets configured)
ERROR: Build failed. Not running target
```

See also: https://github.com/golang/go/issues/57051

### PR DESCRIPTION
**Pull request by malt3**
**What type of PR is this?**

Bug fix

**What package or component does this PR mostly affect?**

go dependencies

**What does this PR do? Why is it needed?**

golang/x/tools no longer ships the vcs package. Instead, there is a new module that now contains the same code.

**Which issues(s) does this PR fix?**

Fixes #1579

**Other notes for review**

### CODE REVIEW

**Comment by malt3** [comment]
I do not understand why gazelle to modify the build file in this way.
We clearly depend on `golang.org/x/tools/go/vcs` (and we even depend on a newer version of `golang.org/x/tools` that doesn't have the vcs package anymore).

```shell-session
$ bazel run //:gazelle_ci 
INFO: Analyzed target //:gazelle_ci (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:gazelle_ci up-to-date:
  bazel-bin/gazelle_ci-runner.bash
  bazel-bin/gazelle_ci
INFO: Elapsed time: 0.106s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/gazelle_ci
--- cmd/fetch_repo/BUILD.bazel  1970-01-01 00:00:00.000000001 +0000
+++ cmd/fetch_repo/BUILD.bazel  1970-01-01 00:00:00.000000001 +0000
@@ -9,7 +9,7 @@
     ],
     importpath = "github.com/bazelbuild/bazel-gazelle/cmd/fetch_repo",
     visibility = ["//visibility:private"],
-    deps = ["@org_golang_x_tools_go_vcs//:vcs"],
+    deps = ["@org_golang_x_tools//go/vcs"],
 )
 
 go_binary(
@@ -22,7 +22,7 @@
     name = "fetch_repo_test",
     srcs = ["fetch_repo_test.go"],
     embed = [":fetch_repo_lib"],
-    deps = ["@org_golang_x_tools_go_vcs//:vcs"],
+    deps = ["@org_golang_x_tools//go/vcs"],
 )
 
 filegroup(
--- language/go/BUILD.bazel     1970-01-01 00:00:00.000000001 +0000
+++ language/go/BUILD.bazel     1970-01-01 00:00:00.000000001 +0000
@@ -84,7 +84,7 @@
         "@com_github_bazelbuild_buildtools//build:go_default_library",
         "@com_github_google_go_cmp//cmp",
         "@io_bazel_rules_go//go/tools/bazel:go_default_library",
-        "@org_golang_x_tools_go_vcs//:vcs",
+        "@org_golang_x_tools//go/vcs",
     ],
 )
 
--- repo/BUILD.bazel    1970-01-01 00:00:00.000000001 +0000
+++ repo/BUILD.bazel    1970-01-01 00:00:00.000000001 +0000
@@ -13,7 +13,7 @@
         "//pathtools",
         "//rule",
         "@org_golang_x_mod//modfile",
-        "@org_golang_x_tools_go_vcs//:vcs",
+        "@org_golang_x_tools//go/vcs",
     ],
 )
 
@@ -29,7 +29,7 @@
         "//pathtools",
         "//rule",
         "//testtools",
-        "@org_golang_x_tools_go_vcs//:vcs",
+        "@org_golang_x_tools//go/vcs",
     ],
 )
 
```

File: MODULE.bazel [hunk 1]
```diff
@@ -4,7 +4,7 @@ module(
     repo_name = "bazel_gazelle",
 )
 
-bazel_dep(name = "bazel_skylib", version = "1.3.0")
```
Comment (fmeum) [inline MODULE.bazel:7]: Is this update needed?

File: deps.bzl [hunk 2]
```diff
@@ -44,10 +44,10 @@ def gazelle_dependencies(
         http_archive,
         name = "bazel_skylib",
         urls = [
-            "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.3.0/bazel-skylib-1.3.0.tar.gz",
```
Comment (fmeum) [inline deps.bzl:47]: Same here, is this needed?

File: internal/bzlmod/default_gazelle_overrides.bzl [hunk 3]
```diff
@@ -7,6 +7,9 @@ DEFAULT_BUILD_FILE_GENERATION_BY_PATH = {
 }
 
 DEFAULT_DIRECTIVES_BY_PATH = {
+    "github.com/bazelbuild/buildtools": [
```
Comment (fmeum) [inline internal/bzlmod/default_gazelle_overrides.bzl:10]: This shouldn't be necessary anymore if you update the `buildtools` dependency to include https://github.com/bazelbuild/buildtools/commit/4f439869c56f3687808fb3d39ebea2e2ca974c7d.

**Review by fmeum — APPROVED** [review]

**Comment by fmeum** [comment]
@malt3 You probably need to add a `# gazelle:repository` directive to the very end of `WORKSPACE`.

File: MODULE.bazel (hunk 1 above)
Comment (malt3) [reply MODULE.bazel]: Not needed. I wasn't sure about warnings regarding minimal version selection with bzlmod.

File: internal/bzlmod/default_gazelle_overrides.bzl (hunk 3 above)
Comment (malt3) [reply internal/bzlmod/default_gazelle_overrides.bzl]: Thanks! Took me a moment to understand how to resolve but I think now the `bzlmod` world and the WORKSPACE world should have the same view of buildtools.

**Review by fmeum — APPROVED** [review]

File: deps.bzl [hunk 4]
```diff
@@ -334,8 +334,15 @@ def gazelle_dependencies(
         go_repository,
         name = "org_golang_x_tools",
         importpath = "golang.org/x/tools",
-        sum = "h1:8WMNJAz3zrtPmnYC7ISf5dEn3MT0gY7jBJfw27yrrLo=",
-        version = "v0.9.1",
+        sum = "h1:YW6HUoUmYBpwSgyaGaZq1fHjrBjX1rlpZ54T6mu2kss=",
+        version = "v0.12.0",
```
Comment (fmeum) [inline deps.bzl:338]: Could you revert this update (also in go.sum)? We should move it into a separate PR.

**Review by fmeum — APPROVED** [review]

File: WORKSPACE [hunk 5]
```diff
@@ -52,7 +52,7 @@ load("@bazel_skylib//lib:unittest.bzl", "register_unittest_toolchains")
 register_unittest_toolchains()
 
 # gazelle:repository go_repository name=co_honnef_go_tools importpath=honnef.co/go/tools
-# gazelle:repository go_repository name=com_github_bazelbuild_buildtools importpath=github.com/bazelbuild/buildtools build_naming_convention=go_default_library
+# gazelle:repository go_repository name=com_github_bazelbuild_buildtools importpath=github.com/bazelbuild/buildtools
```
Comment (fmeum) [inline WORKSPACE:55]: Did you update this to the latest version including the alias change? If it doesn't work, just remove this update and revert this line and we handle it separately.

File: deps.bzl (hunk 4 above)
Comment (malt3) [reply deps.bzl:338]: Will do!

File: WORKSPACE (hunk 5 above)
Comment (malt3) [reply WORKSPACE:55]: The BCR test explicitly downgraded buildtools. This is why the test was red.

**Comment by tyler-french** [comment]
I tested this at Uber and it works (with a patch to buildtools). Could you update `buildtools` again to include https://github.com/bazelbuild/buildtools/commit/691e01a5f313de7b0362cf93914748eea47d0809 ?

Separately and optionally: I took a different approach in https://github.com/bazelbuild/bazel-gazelle/pull/1615 (which has similar logic) to include a ghost import to `x/tools`: https://github.com/bazelbuild/bazel-gazelle/pull/1615/files#diff-fe5a1af491aaa37acf76bdff0f3bb47ddbca27356b676128e21298a4e58b46cb . This approach is a bit hacky, but what if instead we have a file (called `nogodeps.go`), which is just
```
package nogodeps

import (
	// These deps are required by @rules_go//go/tools/builders/nogo_main.go
	// These ghost imports force them to be to be included in the `go.mod`,
	// and therefore in deps.bzl and MODULE.bzl/go_deps
	_ "golang.org/x/tools/go/analysis"
	_ "golang.org/x/tools/go/gcexportdata"
	_ "golang.org/x/tools/internal/facts"
)
```
By including these deps in Go (rather than just Bazel), it also allows the `bazel run //tools/releaser` to update them automatically.

That being said, it is more hacky and therefore less "correct", which could be a drawback.

@fmeum, @linzhp Do you have an opinion on these approaches?

**Comment by malt3** [comment]
> Separately and optionally: I took a different approach in https://github.com/bazelbuild/bazel-gazelle/pull/1615 (which has similar logic) to include a ghost import to x/tools

I personally would be fine with either solution. If you see a benefit in the other, feel free to close this PR and go with the ghost import :).

**Comment by tyler-french** [comment]
> > Separately and optionally: I took a different approach in #1615 (which has similar logic) to include a ghost import to x/tools
> 
> I personally would be fine with either solution. If you see a benefit in the other, feel free to close this PR and go with the ghost import :). It feels like we should implement [bazelbuild/rules_go#3646](https://github.com/bazelbuild/rules_go/issues/3646) to completely remove this class of problems.

Sounds good 😄. Let me know if you think the approach in https://github.com/bazelbuild/bazel-gazelle/pull/1615 is sufficient. One benefit here is that the releaser will automatically update this dep for us, but this is the only real benefit I can think of.
